### PR TITLE
Storagefill component refactor

### LINKED ISSUE #1978: Toolbox contents are defined via component rather than yml
This probably applies to most container entities. Instead of using a custom component to fill the entity, there should be the ability to define their contents in yml using some sort of generic ContainerFillerComponent or somesuch.

If there's a need for some degree of randomizing the contents of a toolbox (ie. electrical toolboxes) then the ContainerFillerComponent could be given that ability too, so you can handle that in yml

### LINKED ISSUE #2046: Randomize StorageFill contents in YAML
In response to #1978 

This is something that should be done in YAML. Having multiple CS files solely dedicated to the fill contents of a single storage container is weird and clunky.

### PR DESCRIPTION
**Pull request by PaulRitter**
closes #1978
closes #2046

StorageFill-Components "contents" list now takes the following arguments

- name (Prototypename of spawnable)
- amount [default: 1] (Amount of items to spawn)
- prob [default: 1f] (probability of the item being spawned)
- orGroup [default: null] (Group of dataset)

The Component will only allow one of all items with the same `orGroup `-parameter to spawn, enabling you to add conditional spawns (see `YellowToolboxItemFilled`)

all standing implementations of StorageFill have been adjusted accordingly of course. Custom Components for StorageFilling have been removed.

**Emergencyclosets no longer have a 1% chance to delete themselves, since thats not possible with my system**

Tested

### CODE REVIEW

**Comment by PaulRitter** [comment]
i fucked up, my head was massivly outdated.

i'll have to update the yaml files & delete fillcomponents. DNM until then

**Comment by RemieRichards** [comment]
I'd prefer if `group` was a bit clearer in its naming, such as `orGroup`

**Comment by RemieRichards** [comment]
also you could name the groups better than 'group1', for the insulated gloves example the desired syntax would be
`orGroup: InsulatedGlovesOrCable`

**Comment by PaulRitter** [comment]
thats actually a good point, i disliked the name too

**Comment by RemieRichards** [comment]
Also on the topic of the PR in general, it's a vast improvement over ad-hoc `XFillComponents`
So big +1 from me.

**Comment by nuke-makes-games** [comment]
upvoted, liked and subscribed

░░░░░░░░░░░░▄▄                     
░░░░░░░░░░░█░░█                  
░░░░░░░░░░░█░░█                  
░░░░░░░░░░█░░░█                  
░░░░░░░░░█░░░░█                  
███████▄▄█░░░░░██████▄ 
▓▓▓▓▓▓█░░░░░░░░░░░░░░█ 
▓▓▓▓▓▓█░░░░ Zumorica, ░░░░░█ 
 ▓▓▓▓▓▓█░░░░Add This░░░░█  
▓▓▓▓▓▓█░░░░Please!░░░░░█ 
▓▓▓▓▓▓█░░░░░░░░░░░░░░█ 
▓▓▓▓▓▓█████░░░░░░░░░█   
██████▀░░░░▀▀██████▀

File: Content.Server/GameObjects/Components/Items/Storage/StorageFillComponent.cs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+                    continue;
+                }
+
+                Spawn(storageItem.PrototypeName);
+                if(storageItem.GroupId != null) alreadySpawnedGroups.Add(storageItem.GroupId);
+            }
+        }
+
+        [Serializable]
+        protected struct PrototypeItemData : IExposeData
```
Comment (exp111) [inline Content.Server/GameObjects/Components/Items/Storage/StorageFillComponent.cs:65]: VS complains that this is created in a sealed class (StorageFillComp)

File: Content.Server/GameObjects/Components/Items/Storage/StorageFillComponent.cs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+                storage.Insert(_entityManager.SpawnEntity(prototype, Owner.Transform.Coordinates));
+            }
+
+            var alreadySpawnedGroups = new List<string>();
+            foreach (var storageItem in _contents)
+            {
+                if (storageItem.PrototypeName == null) continue;
+                if (storageItem.GroupId != null && alreadySpawnedGroups.Contains(storageItem.GroupId)) continue;
+
+                if (!(Math.Abs(storageItem.SpawnProbability - 1f) < 0.001f) &&
```
Comment (exp111) [inline Content.Server/GameObjects/Components/Items/Storage/StorageFillComponent.cs:53]: It's not really clear what you're trying to do here (checking if the SpawnProb == 1f).
Also maybe test if == 1f is enough. If not please add a comment to explain

File: Content.Server/GameObjects/Components/Items/Storage/StorageFillComponent.cs [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        {
+            if (_contents.Count == 0)
+            {
+                return;
+            }
+
+            var storage = Owner.GetComponent<IStorageComponent>();
+            var random = IoCManager.Resolve<IRobustRandom>();
+
+            void Spawn(string prototype)
```
Comment (exp111) [inline Content.Server/GameObjects/Components/Items/Storage/StorageFillComponent.cs:42]: Can't you just call the one line down where you call Spawn? It's not really necessary to call a new func for that imo.
If it's for explanation you could also just add a comment

File: Content.Server/GameObjects/Components/Items/Storage/StorageFillComponent.cs [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        }
+
+        void IMapInit.MapInit()
+        {
+            if (_contents.Count == 0)
+            {
+                return;
+            }
+
+            var storage = Owner.GetComponent<IStorageComponent>();
```
Comment (exp111) [inline Content.Server/GameObjects/Components/Items/Storage/StorageFillComponent.cs:39]: You could do a TryGetComp and if false then Log into the console that the storage doesn't exist. It probably will result in the same, but leaving a message can maybe help debugging faster.

File: Content.Server/GameObjects/Components/Items/Storage/StorageFillComponent.cs [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+namespace Content.Server.GameObjects.Components.Items.Storage
+{
+    [RegisterComponent]
+    internal sealed class StorageFillComponent : Component, IMapInit
+    {
+        [Dependency] private readonly IEntityManager _entityManager;
+
+        public override string Name => "StorageFill";
+
+        private List<PrototypeItemData> _contents = new List<PrototypeItemData>();
```
Comment (exp111) [inline Content.Server/GameObjects/Components/Items/Storage/StorageFillComponent.cs:24]: This initial setting probably isn't necessary as we will overwrite it anyway in ExposeData.
The only thing that would access that array before ExposeData would be Initialize (which we don't use here).

**Review by exp111 — COMMENTED** [review]
Nevermind.

**Review by exp111 — APPROVED** [review]

**Review by DrSmugleaf — APPROVED** [review]

File: Content.Server/GameObjects/Components/Items/Storage/StorageFillComponent.cs [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+            {
+                Logger.Error($"StorageFillComponent couldn't find any StorageComponent ({Owner})");
+                return;
+            }
+            var random = IoCManager.Resolve<IRobustRandom>();
+
+            var alreadySpawnedGroups = new List<string>();
+            foreach (var storageItem in _contents)
+            {
+                if (storageItem.PrototypeName == null) continue;
```
Comment (Zumorica) [inline Content.Server/GameObjects/Components/Items/Storage/StorageFillComponent.cs:48]: Use `string.IsNullOrEmpty` here

File: Content.Server/GameObjects/Components/Items/Storage/StorageFillComponent.cs [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+                if (storageItem.GroupId != null && alreadySpawnedGroups.Contains(storageItem.GroupId)) continue;
```
Comment (Zumorica) [inline Content.Server/GameObjects/Components/Items/Storage/StorageFillComponent.cs:49]: Use `string.IsNullOrEmpty` here too

File: Content.Server/GameObjects/Components/Items/Storage/StorageFillComponent.cs [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+                if (storageItem.SpawnProbability != 1f &&
+                    !random.Prob(storageItem.SpawnProbability))
+                {
+                    continue;
+                }
+
+                storage.Insert(_entityManager.SpawnEntity(storageItem.PrototypeName, Owner.Transform.Coordinates));
+                if(storageItem.GroupId != null) alreadySpawnedGroups.Add(storageItem.GroupId);
```
Comment (Zumorica) [inline Content.Server/GameObjects/Components/Items/Storage/StorageFillComponent.cs:58]: Use `string.IsNullOrEmpty` here too

**Review by Zumorica — APPROVED** [review]
Just a few nitpicks, otherwise it looks good. Thanks!

**Pull request merged**